cdc/puller/buffer.go

```diff
@@ -37,4 +55,112 @@ func (b Buffer) Get(ctx context.Context) (model.RegionFeedEvent, error) {
 	}
 }
 
-// TODO limit memory buffer
+var _ EventBuffer = &memBuffer{}
+
+type memBuffer struct {
+	limitter *BlurResourceLimitter
+
+	mu struct {
+		sync.Mutex
+		entries deque.Deque
+	}
+	signalCh chan struct{}
+}
+
+// Passing nil will make a unlimited  buffer.
+func makeMemBuffer(limitter *BlurResourceLimitter) *memBuffer {
+	return &memBuffer{
+		limitter: limitter,
+		mu: struct {
+			sync.Mutex
+			entries deque.Deque
+		}{
+			entries: deque.NewDeque(),
+		},
+
+		signalCh: make(chan struct{}, 1),
+	}
+}
+
+// AddEntry implements EventBuffer interface.
+func (b *memBuffer) AddEntry(ctx context.Context, entry model.RegionFeedEvent) error {
+	b.mu.Lock()
+	if b.limitter != nil && b.limitter.OverBucget() {
+		b.mu.Unlock()
+		return ErrReachLimit
+	}
+
+	b.mu.entries.PushBack(entry)
+	if b.limitter != nil {
+		b.limitter.Add(int64(entrySize(entry)))
+	}
+	b.mu.Unlock()
+
+	select {
+	case b.signalCh <- struct{}{}:
+	default:
+	}
+
+	return nil
+}
+
+// Get implements EventBuffer interface.
+func (b *memBuffer) Get(ctx context.Context) (model.RegionFeedEvent, error) {
+	for {
+		b.mu.Lock()
+		if !b.mu.entries.Empty() {
+			e := b.mu.entries.PopFront().(model.RegionFeedEvent)
+			if b.limitter != nil {
+				b.limitter.Add(int64(-entrySize(e)))
+			}
+			b.mu.Unlock()
+			return e, nil
+		}
+
+		b.mu.Unlock()
+
+		select {
+		case <-ctx.Done():
+			return model.RegionFeedEvent{}, ctx.Err()
+		case <-b.signalCh:
+		}
+	}
+}
+
+var sizeOfVal = unsafe.Sizeof(model.RawKVEntry{})
+var sizeOfResolve = unsafe.Sizeof(model.ResolvedSpan{})
+
+func entrySize(e model.RegionFeedEvent) int {
+	if e.Val != nil {
+		return int(sizeOfVal) + len(e.Val.Key) + len(e.Val.Value)
+	} else if e.Resolved != nil {
+		return int(sizeOfResolve)
+	} else {
+		log.Fatal("unknow event type")
+	}
+
+	return 0
+}
+
+// BlurResourceLimitter limit resource use.
+type BlurResourceLimitter struct {
+	bucget int64
```

Comment:
what does `bucget` mean, is it `budget` or `bucket`

Reply:
budget, fixed